### PR TITLE
[llvm][lld] Pre-commit tests for RISCV TLSDESC symbols

### PR DESCRIPTION
**Pull request by ilovepi**
Currently, we mistakenly mark the local labels used in RISC-V TLSDESC as
TLS symbols, when they should not be. This patch adds tests with the
current incorrect behavior, and subsequent patches will address the
issue.

### CODE REVIEW

File: lld/test/ELF/riscv-tlsdesc.s [hunk 1]
```diff
@@ -29,6 +29,12 @@
 # RUN: ld.lld -e 0 -z now a.32.o c.32.so -o a.32.ie
 # RUN: llvm-objdump --no-show-raw-insn -M no-aliases -h -d a.32.ie | FileCheck %s --check-prefix=IE32
 
+# RUN: llvm-mc -triple=riscv64 --position-independent -filetype=obj d.s -o d.64.o
+# RUN: not ld.lld -shared -soname=libd.64.so -o libd.64.so d.64.o 2>&1 | FileCheck %s --check-prefix=BADTLSLABEL
```
Comment (MaskRay) [inline lld/test/ELF/riscv-tlsdesc.s:33]: we don't use `lib*.so` in tests. Just d.64.so

File: lld/test/ELF/riscv-tlsdesc.s [hunk 2]
```diff
@@ -192,3 +201,31 @@ b:
...
+  .globl	bar
+  .p2align	1
+  .type	bar,@function
+bar:
+  addi	sp, sp, -16
+.Lpcrel_hi0:
+  auipc	a0, %got_pcrel_hi(_ZTH3foo)
+  ld	a0, %pcrel_lo(.Lpcrel_hi0)(a0)
+  beqz	a0, .Ltlsdesc_hi0
+  call	_ZTH3foo
```
Comment (MaskRay) [inline lld/test/ELF/riscv-tlsdesc.s:218]: remove unneeded instructions from the test

File: llvm/test/CodeGen/RISCV/tlsdesc-symbol.ll [hunk 3]
```diff
@@ -0,0 +1,22 @@
+; RUN: llc -mtriple=riscv64 -relocation-model=pic -enable-tlsdesc < %s \
+; RUN:     | llvm-mc -triple=riscv64 -filetype=obj -o %t.o --position-independent \
```
Comment (MaskRay) [inline llvm/test/CodeGen/RISCV/tlsdesc-symbol.ll:2]: If `-o %t.o`, the pipe should not be used

Comment (MaskRay) [reply llvm/test/CodeGen/RISCV/tlsdesc-symbol.ll:2]: I don't think a new file is needed. You can additionally add `llvm-readelf -s` lines to tls-models.ll

Comment (ilovepi) [reply llvm/test/CodeGen/RISCV/tlsdesc-symbol.ll:2]: I used a separate file since the other checks were autogenerated.

Comment (ilovepi) [reply llvm/test/CodeGen/RISCV/tlsdesc-symbol.ll]: I think the latest version now avoids the output file. Thanks for catching that.

File: lld/test/ELF/riscv-tlsdesc.s (hunk 2 above)
Comment (ilovepi) [reply lld/test/ELF/riscv-tlsdesc.s]: Thanks. The new version should be much smaller, but let me know if I should reduce it more.

File: lld/test/ELF/riscv-tlsdesc.s (hunk 1 above)
Comment (ilovepi) [reply lld/test/ELF/riscv-tlsdesc.s]: done.

File: lld/test/ELF/riscv-tlsdesc.s [hunk 4]
```diff
@@ -29,6 +29,12 @@
 # RUN: ld.lld -e 0 -z now a.32.o c.32.so -o a.32.ie
 # RUN: llvm-objdump --no-show-raw-insn -M no-aliases -h -d a.32.ie | FileCheck %s --check-prefix=IE32
 
+# RUN: llvm-mc -triple=riscv64 --position-independent -filetype=obj d.s -o d.64.o
```
Comment (MaskRay) [inline lld/test/ELF/riscv-tlsdesc.s:32]: Delete --position-independent, which is only relevant for mips.

**Review by MaskRay — APPROVED** [review]

File: llvm/test/CodeGen/RISCV/tlsdesc-symbol.ll (hunk 3 above)
Comment (ilovepi) [reply llvm/test/CodeGen/RISCV/tlsdesc-symbol.ll]: Just to confirm, we're OK w/ the extra test file? or shall I add manual tests to `tls-model.ll`?

File: lld/test/ELF/riscv-tlsdesc.s (hunk 4 above)
Comment (ilovepi) [reply lld/test/ELF/riscv-tlsdesc.s]: Thanks for pointing that out.

File: llvm/test/CodeGen/RISCV/tlsdesc-symbol.ll [hunk 5]
```diff
@@ -0,0 +1,22 @@
+; RUN: llc -mtriple=riscv64 -relocation-model=pic -enable-tlsdesc < %s \
```
Comment (MaskRay) [inline llvm/test/CodeGen/RISCV/tlsdesc-symbol.ll:1]: Yes, this extra file is file.

Best to reference tls-models.ll

**Comment by ilovepi** [comment]
\cherry-pick #85816

**Pull request merged**